src/Microsoft.Framework.DependencyInjection.Abstractions/ServiceCollectionExtensions.cs

```diff
@@ -184,6 +185,53 @@ public static bool TryAddSingleton<TService>([NotNull] this IServiceCollection c
             return TryAddSingleton(collection, typeof(TService), typeof(TImplementation));
         }
 
+        /// <summary>
+        /// Adds a <see cref="ServiceDescriptor"/> if an existing descriptor with the same
+        /// <see cref="ServiceDescriptor.ServiceType"/> and <see cref="ServiceDescriptor.ImplementationType"/> does
+        /// not already exist in <paramref name="services."/>.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="descriptor">
+        /// A <see cref="ServiceDescriptor"/> with <see cref="ServiceDescriptor.ImplementationType"/> set to a
+        /// non-<c>null</c> value.
+        /// </param>
+        /// <returns>
+        /// <c>true</c> if <paramref name="descriptor"/> is added to the collection, otherwise <c>false</c>.
+        /// </returns>
+        /// <remarks>
+        /// Use <see cref="TryAddMultiRegistration"/> when registing a service implementation of a service type that
+        /// supports multiple registrations of the same service type. Using
+        /// <see cref="Add(IServiceCollection, ServiceDescriptor)"/> is not idempotent and can add duplicate
+        /// <see cref="ServiceDescriptor"/> instances if called twice. Using
+        /// <see cref="TryAddMultiRegistration(IServiceCollection, ServiceDescriptor)"/> will prevent registration
+        /// of multiple implementation types.
+        /// </remarks>
+        public static bool TryAddMultiRegistration(
```

Comment:
Naming seems weird since we don't have a real concept of adding to an IEnumerable<TService> distinguished from replacing a previous definition of a service type.